### PR TITLE
Switch to spotbugs and fix some issues

### PR DESCRIPTION
**Pull request by res0nance**
- Swapped to spotbugs and fixed multiple issues.
- There still remains some issues but the PR was getting large

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
Need to review it carefully, but thanks a lot!

**Comment by oleg-nenashev** [comment]
sorry, still in my queue

File: src/main/java/winstone/Ajp13ConnectorFactory.java [hunk 1]
```diff
@@ -15,14 +15,14 @@
...
- * 
+ *
  * @author mailto: <a href="rick_knowles@hotmail.com">Rick Knowles</a>
  * @version $Id: Ajp13ConnectorFactory.java,v 1.12 2006/03/24 17:24:22 rickknowles Exp $
  */
 public class Ajp13ConnectorFactory implements ConnectorFactory {
     public boolean start(Map args, Server server) throws IOException {
         int listenPort = Option.AJP13_PORT.get(args);
-        String listenAddress = Option.AJP13_LISTEN_ADDRESS.get(args);
+        // String listenAddress = Option.AJP13_LISTEN_ADDRESS.get(args);
```
Comment (oleg-nenashev) [inline src/main/java/winstone/Ajp13ConnectorFactory.java]: If unused, please.do one of the following..

* delete it
* Create a follow-up bug, link it from a TODO comment

Looks like it is a bug unless I miss something

File: src/main/java/winstone/HostConfiguration.java [hunk 2]
```diff
@@ -250,43 +252,51 @@ protected File getWebRoot(File requestedWebroot, File warfile) throws IOExceptio
...
-                    while ( readBytes != -1 ) {
-                        outStream.write( buffer, 0, readBytes );
-                        readBytes = inContent.read( buffer );
+                    // If archive date is newer than unzipped file, overwrite
+                    File outFile = new File(unzippedDir, elemName);
+                    if (outFile.exists() && (outFile.lastModified() > warfile.lastModified())) {
+                        continue;
+                    }
+                    File parentOutFile = outFile.getParentFile();
+                    if (!parentOutFile.mkdirs()) {
```
Comment (oleg-nenashev) [inline src/main/java/winstone/HostConfiguration.java]: We can drop Java 6 later and start using `Files` here.
CC @olamy

File: src/main/java/winstone/realm/FileRealm.java [hunk 3]
```diff
@@ -34,12 +34,12 @@
...
 public class FileRealm extends HashLoginService {
     private static final WinstoneResourceBundle REALM_RESOURCES = new WinstoneResourceBundle("winstone.realm.LocalStrings");
-    
-    final String DEFAULT_FILE_NAME = "users.xml";
-    final String ELEM_USER = "user";
-    final String ATT_USERNAME = "username";
-    final String ATT_PASSWORD = "password";
-    final String ATT_ROLELIST = "roles";
+
+    static final String DEFAULT_FILE_NAME = "users.xml";
```
Comment (oleg-nenashev) [inline src/main/java/winstone/realm/FileRealm.java]: Breaks public API?

Comment (res0nance) [reply src/main/java/winstone/realm/FileRealm.java]: I don;t see how you should still be able to access the variable via a static or nonstatic way

File: src/main/java/winstone/HostConfiguration.java (hunk 2 above)
Comment (res0nance) [reply src/main/java/winstone/HostConfiguration.java]: I'll clean this bit up then

File: src/main/java/winstone/realm/FileRealm.java (hunk 3 above)
Comment (oleg-nenashev) [reply src/main/java/winstone/realm/FileRealm.java]: Right, sorry

File: src/main/java/winstone/tools/WinstoneControl.java [hunk 4]
```diff
@@ -22,20 +22,20 @@
...
-    
+
     final static String OPERATION_SHUTDOWN = "shutdown";
     final static String OPERATION_RELOAD = "reload:";
     static int TIMEOUT = 10000;
 
-    public static OInt CONTROL_PORT = Option.integer("controlPort");
-    public static OInt PORT = Option.integer("port");
-    public static OInt DEBUG = new OInt("debug", 5) {
+    public final static OInt CONTROL_PORT = Option.integer("controlPort");
```
Comment (oleg-nenashev) [inline src/main/java/winstone/tools/WinstoneControl.java]: IIRC breaking change was here. Not that important

File: src/main/java/winstone/HostConfiguration.java [hunk 5]
```diff
@@ -230,10 +232,12 @@ protected File getWebRoot(File requestedWebroot, File warfile) throws IOExceptio
                 File tempFile = File.createTempFile("dummy", "dummy");
                 String userName = System.getProperty("user.name");
                 unzippedDir = new File(tempFile.getParent(),
-                        (userName != null ? WinstoneResourceBundle.globalReplace(userName, 
+                        (userName != null ? WinstoneResourceBundle.globalReplace(userName,
                                 new String[][] {{"/", ""}, {"\\", ""}, {",", ""}}) + "/" : "") +
                         "winstone/" + warfile.getName());
-                tempFile.delete();
+                if (!tempFile.delete()) {
```
Comment (jglick) [inline src/main/java/winstone/HostConfiguration.java]: Or there is a Commons IO utility, or you can use `java.nio.file.Files.delete` which throws a meaningful exception.

File: src/main/java/winstone/HostConfiguration.java [hunk 6]
```diff
@@ -250,43 +254,52 @@ protected File getWebRoot(File requestedWebroot, File warfile) throws IOExceptio
             if(!timestampFile.exists() || Math.abs(timestampFile.lastModified()- warfile.lastModified())>1000) {
                 // contents of the target directory is inconsistent from the war.
                 deleteRecursive(unzippedDir);
-                unzippedDir.mkdirs();
+                if (!unzippedDir.mkdirs()) {
```
Comment (jglick) [inline src/main/java/winstone/HostConfiguration.java]: ditto

File: src/main/java/winstone/HostConfiguration.java [hunk 7]
```diff
@@ -250,43 +254,52 @@ protected File getWebRoot(File requestedWebroot, File warfile) throws IOExceptio
...
-                    while ( readBytes != -1 ) {
-                        outStream.write( buffer, 0, readBytes );
-                        readBytes = inContent.read( buffer );
+                    // If archive date is newer than unzipped file, overwrite
+                    File outFile = new File(unzippedDir, elemName);
+                    if (outFile.exists() && (outFile.lastModified() > warfile.lastModified())) {
+                        continue;
+                    }
+                    try {
+                        Files.createDirectories(Paths.get(outFile.getParent()));
```
Comment (jglick) [inline src/main/java/winstone/HostConfiguration.java]: FWIW in Jenkins core we have been catching `InvalidPathException` as well, which apparently can happen under some obscure conditions on Windows servers.

Comment (oleg-nenashev) [reply src/main/java/winstone/HostConfiguration.java]: Would recommend to address it before merging

**Pull request merged**